src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -799,516 +899,41 @@ func (s *commitLogSource) shouldEncodeForData(
-			hasSnapshotBlock bool
-		)
-
-		if snapshotData.Blocks != nil {
-			snapshotBlock, hasSnapshotBlock = snapshotData.Blocks.BlockAt(start)
-		}
-
-		if !hasSnapshotBlock {
-			// Make sure snapshotBlock is nil if it does not exist.
-			snapshotBlock = nil
-		}
-
-		// Closes encoders and snapshotBlock by calling Discard() on each.
-		readers, err := newIOReadersFromEncodersAndBlock(
-			segmentReaderPool, encoders, snapshotBlock)
-		if err != nil {
-			numErrs++
-			continue
-		}
-
-		iter := multiReaderIteratorPool.Get()
-		iter.Reset(readers, time.Time{}, 0, nsCtx.Schema)
-
-		enc := encoderPool.Get()
-		enc.Reset(start, blopts.DatabaseBlockAllocSize(), nsCtx.Schema)
-		for iter.Next() {
-			dp, unit, annotation := iter.Current()
-			encodeErr := enc.Encode(dp, unit, annotation)
-			if encodeErr != nil {
-				err = encodeErr
-				numErrs++
-				break
-			}
-		}
-
-		if iterErr := iter.Err(); iterErr != nil {
-			if err == nil {
-				err = iter.Err()
-			}
-			numErrs++
-		}
-
-		// Automatically returns iter to the pool
-		iter.Close()
-		readers.close()
-		if hasSnapshotBlock {
-			// Block is already closed, but we need to remove from the Blocks
-			// to prevent a double free when we call Blocks.Close() later.
-			snapshotData.Blocks.RemoveBlockAt(start)
-		}
-
-		if err != nil {
-			continue
-		}
-
-		pooledBlock := blocksPool.Get()
-		pooledBlock.Reset(start, blockSize, enc.Discard(), nsCtx)
-		if seriesBlocks == nil {
-			seriesBlocks = block.NewDatabaseSeriesBlocks(len(unmergedCommitlogBlocks.encoders))
-		}
-		seriesBlocks.AddBlock(pooledBlock)
-	}
-
-	if snapshotData.Blocks != nil {
-		allSnapshotBlocks := snapshotData.Blocks.AllBlocks()
-		for startNano, snapshotBlock := range snapshotData.Blocks.AllBlocks() {
-			if seriesBlocks == nil {
-				seriesBlocks = block.NewDatabaseSeriesBlocks(len(allSnapshotBlocks))
-			}
-			_, ok := seriesBlocks.BlockAt(startNano.ToTime())
-			if ok {
-				// Should never happen because we would have called
-				// Blocks.RemoveBlockAt() above.
-				iOpts := s.opts.CommitLogOptions().InstrumentOptions()
-				instrument.EmitAndLogInvariantViolation(iOpts, func(l *zap.Logger) {
-					l.Error("tried to merge block that should have been removed",
-						zap.Time("blockStart", startNano.ToTime()))
-				})
-				continue
-			}
-
-			seriesBlocks.AddBlock(snapshotBlock)
-		}
-	}
-	return seriesBlocks, numEmptyErrs, numErrs
-}
-
-func (s *commitLogSource) findHighestShard(shardsTimeRanges result.ShardTimeRanges) uint32 {
-	var max uint32
-	for shard := range shardsTimeRanges {
-		if shard > max {
-			max = shard
-		}
-	}
-	return max
-}
-
-func (s *commitLogSource) logEncodingOutcome(workerErrs []int, iter commitlog.Iterator) {
-	errSum := 0
-	for _, numErrs := range workerErrs {
-		errSum += numErrs
+func (s *commitLogSource) logAccumulateOutcome(
+	workers []*accumulateWorker,
+	iter commitlog.Iterator,
+) {
+	errs := 0
+	for _, worker := range workers {
+		errs += worker.numErrors
```

Comment:
Might be useful to also note the number of workers that encountered errors? Just for potential scenarios where there's a worker that got into a corrupt state somehow

Reply:
Hm, I might opt not to do this for now - especially since we just enqueue across all workers without any stickyness or stateful selection of which worker gets to work on what data.